### PR TITLE
assert: build deepStrictEqual api.

### PR DESCRIPTION
**Pull request by Txiaozhe**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,7 +8,8 @@
     "test-jerry": "deps/jerry/tools/run-tests.py --unittests",
     "test-iotjs": "tools/build.py --run-test --no-check-valgrind",
     "build": "tools/build.py",
-    "rebuild": "tools/build.py --clean"
+    "rebuild": "tools/build.py --clean",
+    "myrebuild": "tools/build.py --clean --external-include-dir /usr/local/Cellar/dbus/1.12.10/include/dbus-1.0 --external-include-dir /usr/local/lib/dbus-1.0/include"
```
Comment (yorkie) [inline package.json]: Remove this line

File: src/modules.json [hunk 2]
```diff
@@ -56,7 +56,11 @@
     },
     "assert": {
       "js_file": "js/assert.js",
-      "require": ["util"]
+      "require": ["util", "internal_comparisons"]
+    },
+    "internal_comparisons": {
```
Comment (yorkie) [inline src/modules.json]: Should be `internal_util_comparisions`.

File: src/modules.json [hunk 3]
```diff
@@ -56,7 +56,11 @@
...
+      "js_file": "internal/util/comparisons.js",
+      "require": ["console"]
```
Comment (yorkie) [inline src/modules.json]: This dependents is no need.

File: package.json (hunk 1 above)
Comment (Txiaozhe) [reply package.json]: I am sorry, I push the test file! Awkward!!!

File: src/internal/util/comparisons.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+function isError(err) {
+  return err instanceof Error;
+}
+
+function isArray(arr) {
+  return Array.isArray(arr);
+}
+
+// Temporary use before Number.isNaN() is supported.
+function numberIsNaN(num) {
```
Comment (yorkie) [inline src/internal/util/comparisons.js]: We have the function `isNaN`.

File: src/js/assert.js [hunk 5]
```diff
@@ -99,12 +97,74 @@ function strictEqual(actual, expected, message) {
...
+function strictDeepEqual(val1, val2) {
+  if (util.isNumber(val1)) {
+    if (util.isNumber(val2)) {
+      return (Number.isNaN(val1) && Number.isNaN(val2)) || (val1 === val2);
+    } else {
+      return false;
+    }
+  }
+
+  if (util.isArray(val1)) {
```
Comment (yorkie) [inline src/js/assert.js]: I think we should use Array.isArray directly :)

File: src/js/assert.js [hunk 6]
```diff
@@ -99,12 +97,74 @@ function strictEqual(actual, expected, message) {
...
+    if (util.isArray(val2) && val1.length === val2.length) {
```
Comment (yorkie) [inline src/js/assert.js]: Ditto.

File: src/js/util.js [hunk 7]
```diff
@@ -290,6 +301,8 @@ exports.isFinite = isFinite;
 exports.isFunction = isFunction;
 exports.isBuffer = isBuffer;
 exports.isRegExp = isRegExp;
+exports.isDate = isDate;
+exports.isError = isError;
 exports.isArray = Array.isArray;
```
Comment (Txiaozhe) [inline src/js/util.js]: But why we export it here? Why **Array.isArray** directly?

Comment (Txiaozhe) [inline src/js/util.js]: But why we export it here? Why Array.isArray directly?

Comment (yorkie) [reply src/js/util.js]: This seems to be exported before from IoT.js.

File: src/js/assert.js [hunk 8]
```diff
@@ -97,19 +96,76 @@ function strictEqual(actual, expected, message) {
...
-    Object.keys(actual).length,
-    Object.keys(expected).length,
-    message
-  );
-
-  for (var key in expected) {
-    deepStrictEqual(actual[key], expected[key], message);
+  if (!isDeepStrictEqual(actual, expected)) {
+    fail(actual, expected, message, '===');
   }
```
Comment (yorkie) [inline src/js/assert.js]: @Txiaozhe Could you please run test once locally? You seems changed the behavior of this function.

Comment (Txiaozhe) [reply src/js/assert.js]: I run the `npm test`, get this error: 
```shell
IoT.js Build Succeeded!!

==> Build test dependencies

==> Build NAPI test module with /Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp

/Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp configure

/Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp build

  CC(target) Release/obj.target/napi_construct/napi_construct.o
../napi_construct.c:35:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
NAPI_MODULE_INIT() {
^
../napi_construct.c:37:39: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
                                      ^
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
            ^
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
../napi_construct.c:37:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:39:40: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
                                       ^
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
            ^
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
../napi_construct.c:39:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:40:44: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
                                           ^
../napi_construct.c:40:58: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
                                                         ^
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
            ^
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
../napi_construct.c:40:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:42:20: error: use of undeclared identifier 'env'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                   ^
../napi_construct.c:42:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:65:14: note: expanded from macro 'SET_NAMED_METHOD'
      return NULL;                                                   \
             ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../napi_construct.c:42:20: error: use of undeclared identifier 'env'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                   ^
../napi_construct.c:42:25: error: use of undeclared identifier 'exports'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                        ^
../napi_construct.c:42:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:69:14: note: expanded from macro 'SET_NAMED_METHOD'
      return NULL;                                                   \
             ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../napi_construct.c:43:10: error: use of undeclared identifier 'exports'
  return exports;
         ^
6 warnings and 17 errors generated.
make: *** [Release/obj.target/napi_construct/napi_construct.o] Error 1
```
what happened about **napi**? Awkward so mush!

Comment (legendecas) [reply src/js/assert.js]: These issues may caused by not compatible Node.js version. Try latest LTS version like v10.13 or newer and run tests again.

**Comment by legendecas** [comment]
I'd recommend rebasing current branch onto latest master branch to eliminate unrelated commit to find out what issue caused CI failure.

File: src/js/assert.js (hunk 8 above)
Comment (Txiaozhe) [reply src/js/assert.js]: Maybe I found the error. we have not `isNaN()` function in Number or util?
```js
'use strict';

var util = require('util');

console.log(util.isNaN(3));
console.log(Number.isNaN(3));
```
output:
```shell
TypeError: Expected a function.
    at undefined (/Users/jimtang/GitHub/ShadowNode/number.js:1:1)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:266:36)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:214:23)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:301:26)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/iotjs.js.wrapped:18:2)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/iotjs.js.wrapped:1:1)
```

**Comment by Txiaozhe** [comment]
> I'd recommend rebasing current branch onto latest master branch to eliminate unrelated commit to find out what issue caused CI failure.

Thanks! Maybe I will do.

**Comment by Txiaozhe** [comment]
> > I'd recommend rebasing current branch onto latest master branch to eliminate unrelated commit to find out what issue caused CI failure.
> 
> Thanks! Maybe I will do.

I found the error, because of we have not isNaN function!

**Comment by Txiaozhe** [comment]
when I rebase it on master, it will be a conflict about **mqtt**, It is not very  suitable.

**Comment by Txiaozhe** [comment]
It can be merged directly?

File: src/js/util.js [hunk 9]
```diff
@@ -5,6 +5,11 @@ function isNull(arg) {
 }
 
 
+function isNaN(arg) {
+  return String(arg) === 'NaN';
```
[the remark this answers is not in the document]
Comment (Txiaozhe) [reply src/js/util.js]: Indeed, it seems even more strictly, but this operate will surplus mount a function at 'Number' object, but in my submission, we should implement a real 'isNaN' function in 'Number' or our 'util' module but not do this. 

```js
function isNaN(arg) {
  return String(arg) === 'NaN';
}
```
May be my this function is temporary.

[the remark this answers is not in the document]
Comment (legendecas) [reply src/js/util.js]: A global `isNaN` does exists in ShadowNode.
```
isNaN(233) // => false
```

**Comment by Txiaozhe** [comment]
Thanks, for convenience, I will redo it! @yorkie @legendecas

**Pull request closed without merging**